src/adapter/rust_type_name.rs

```diff
@@ -0,0 +1,676 @@
+            if !fnp.generic_params.is_empty() {
+                write!(f, "for<")?;
+                intersperse_with(f, ", ", &fnp.generic_params, GenericParamDef)?;
+                write!(f, "> ")?;
+            }
+
+            if fnp.header.is_const {
+                write!(f, "const ")?;
+            }
+
+            if fnp.header.is_async {
+                write!(f, "async ")?;
+            }
+
+            if fnp.header.is_unsafe {
+                write!(f, "unsafe ")?;
+            }
+
+            macro_rules! abi_name {
+                ($f:expr, $name:literal, $unwind:expr) => {{
+                    write!($f, concat!("extern \"", $name))?;
+
+                    if *$unwind {
+                        write!(f, "-unwind")?;
+                    }
+
+                    write!(f, "\" ")?;
+                }};
+            }
+
+            match &fnp.header.abi {
+                rustdoc_types::Abi::Rust => (),
+                rustdoc_types::Abi::C { unwind } => abi_name!(f, "C", unwind),
+                rustdoc_types::Abi::Cdecl { unwind } => abi_name!(f, "cdecl", unwind),
+                rustdoc_types::Abi::Stdcall { unwind } => abi_name!(f, "stdcall", unwind),
+                rustdoc_types::Abi::Fastcall { unwind } => abi_name!(f, "fastcall", unwind),
+                rustdoc_types::Abi::Aapcs { unwind } => abi_name!(f, "aapcs", unwind),
+                rustdoc_types::Abi::Win64 { unwind } => abi_name!(f, "win64", unwind),
+                rustdoc_types::Abi::SysV64 { unwind } => abi_name!(f, "sysv64", unwind),
+                rustdoc_types::Abi::System { unwind } => abi_name!(f, "system", unwind),
+                rustdoc_types::Abi::Other(other) => write!(f, r#"extern "{other}" "#)?,
+            }
+
+            write!(f, "fn(")?;
+
+            enum Arg<'a> {
+                Named(&'a str, Type<'a>),
+                Dots,
+            }
+
+            intersperse(
+                f,
+                ", ",
+                fnp.sig
+                    .inputs
+                    .iter()
+                    .map(|(name, ty)| Arg::Named(name, Type(ty)))
+                    .chain(fnp.sig.is_c_variadic.then_some(Arg::Dots)),
+                |arg, f| match arg {
+                    Arg::Named(name, ty) => write!(f, "{name}: {ty}"),
+                    Arg::Dots => write!(f, "..."),
+                },
+            )?;
+
+            write!(f, ")")?;
+
+            if let Some(output) = &fnp.sig.output {
+                write!(f, " -> {}", Type(output))?;
+            }
+
+            Ok(())
+        }
+        rustdoc_types::Type::Tuple(tys) => {
+            write!(f, "(")?;
+            intersperse_with(f, ", ", tys, Type)?;
+            write!(f, ")")
+        }
+        rustdoc_types::Type::Slice(ty) => write!(f, "[{}]", Type(ty)),
+        // According to the docs for this variant, `len` is not guaranteed to be the
+        // same as the source code.  For example, [u8; 1 + 2] currently becomes [u8; 3]
```

Comment:
```suggestion
        // According to the docs for this variant, `len` is not guaranteed to be the
        // same as the source code.  For example, `[u8; 1 + 2]` currently becomes `[u8; 3]`.
```
Does this appear to be guaranteed to have the actual constant value present? E.g. (and I'm not sure if this is legal Rust) if we have `[u8; some_const_fn() + 1]` or something like that.

Or can it ever be `_` like the values of `const` items sometimes end up being in cases like `const ITEM: u8 = some_const_fn();` ?

Reply:
The docstring just says 

> Keep in mind that it's not guaranteed to match the actual source code of the expression.

```rust
const fn const_fn() -> usize {
    let mut i = 0;

    while i < 27 {
        i += 5;
    }

    i
}

// ...
    pub array: [(); 1 + 2],
    pub array2: [(); usize::max >> (usize::bits - 1)],
    pub array3: [(); const_fn()],
```

is serialized as:

```rs
array: [(); 3],
array2: [(); 1],
array3: [(); 30]
```

I don't know if there are any guarantees, but I added these test cases.

Reply:
Very nice. Having the test cases is great, so we'll know if this behavior changes in the future. Thanks!